cubeb-core/src/ffi_types.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright © 2017-2018 Mozilla Foundation
+//
+// This program is made available under an ISC-style license.  See the
+// accompanying file LICENSE for details.
+
+/// A macro to define a wrapper around a POD FFI type that lives on the
+/// heap.  This differs from `foreign_type!` because that macros assumes
+/// that drop can be implemented just by passing the pointer to a
+/// function. Cubeb's APIs are more complicated.
```

Comment:
Maybe I need to compare this side-by-side with the foreign_types version, but explaining how the drop stuff is different would be good.

Reply:
So the `foreign_type!` macro must be given `fn drop = ...` and the `impl Drop` for the struct is implemented to call the function, passing the wrapped pointer, ie:
```rust
extern pub unsafe fn cubeb_destroy(*mut ffi::cubeb);

foreign_type! {
    type CType = ffi::cubeb;
    fn drop = cubeb_destroy;
    ...
}
```

This won't work for `cubeb_stream_device_destroy` which takes two parameters.  (Maybe all the cubeb destroy interfaces should be changed to just take a pointer to the object that is being destroyed)

I'll update the comment.

Reply:
Thanks, makes sense now.